Add unique id for todoist calendar entity

boralyl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This ticket addresses #75509 and adds a unique id for todoist calendar entities so that they can be managed from the UI.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #75509 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

boralyl committed:
Add unique id for todoist calendar entity.

boralyl commented on homeassistant/components/todoist/calendar.py:
```diff
@@ -296,6 +296,7 @@ def __init__(
         )
         self._cal_data = {}
         self._name = data[CONF_NAME]
+        self._attr_unique_id = data[ID]
```
This is the ID of the calendar as returned by the API and is stable.

MartinHjelmare replied:
I don't see this key in the data passed to the entity.

Please also add a test.

boralyl replied:
It's not obvious from this diff, but it happens here: https://github.com/home-assistant/core/blob/fcbe2cc44be0172153a272fa43e72eb18bb3f790/homeassistant/components/todoist/calendar.py#L140-L141

I switched to the constant that is being used above, even though they are the same values for consistency.

MartinHjelmare replied:
And how does it work for the extra projects?

https://github.com/home-assistant/core/blob/0df08b6b0ce6e8d853ba4981e8f1e1492227a198/homeassistant/components/todoist/calendar.py#L169-L180

boralyl replied:
That's a great point!  Thanks for noticing that. Since the user configures these via yaml, it's probably safe to return None for those entities `unique_id`. I updated setting the unique id attribute to take that into account.

MartinHjelmare reviewed (changes requested):
See comment above.

boralyl committed:
Add unit test.

boralyl committed:
Use same constant as used in setup function.

MartinHjelmare commented on tests/components/todoist/test_calendar.py:
```diff
@@ -42,3 +46,14 @@ def test_parse_due_date_without_timezone_uses_offset():
+
+def test_calendar_entity_unique_id():
+    """Test unique id is set to project id."""
+    entity = TodoistProjectEntity(
```
Please set up the integration, while patching the library appropriately to avoid I/O, and get the entity entry from the entity registry and assert its state.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

boralyl replied:
Thank you, I somehow had never seen those docs, likely because this component is mostly untested.

boralyl committed:
Update test to conform to standards.

Also because custom projects will not have an ID, use .get when
retrieving the id as it may not be present.

boralyl commented on homeassistant/components/todoist/calendar.py:
```diff
@@ -296,6 +296,7 @@ def __init__(
         )
         self._cal_data = {}
         self._name = data[CONF_NAME]
+        self._attr_unique_id = data.get(CONF_ID)
```
If this is a custom project configured in yaml by the user the unique_id will be `None`, otherwise it will return the stable ID of the project from the api.

MartinHjelmare replied:
Good.

boralyl committed:
Remove unused config key

boralyl commented on tests/components/todoist/test_calendar.py:
```diff
@@ -42,3 +49,36 @@ def test_parse_due_date_without_timezone_uses_offset():
+
+
+@patch("homeassistant.components.todoist.calendar.TodoistAPI")
+async def test_calendar_entity_unique_id(todoist_api, hass, state):
```
I attempted to write a test for the custom project case to test that it's unique_id is `None` but that entity never shows up in the entity registry.  I can look up its state via `hass.states`, but it wasn't in the registry.  I might be missing awaiting something, but I couldn't figure that out.

MartinHjelmare replied:
If the entity doesn't have a unique id it won't be part of the entity registry. Assert that that's true.

boralyl replied:
Oh, that makes sense, thank you!

boralyl committed:
Add test for custom project calendar.

MartinHjelmare commented on tests/components/todoist/test_calendar.py:
```diff
@@ -81,3 +81,26 @@ async def test_calendar_entity_unique_id(todoist_api, hass, state):
+
+    registry = entity_registry.async_get(hass)
+    entity = registry.async_get("calendar.all_projects")
+    assert entity is None
```
Please also assert that there's a state for this custom calendar entity. Look up `hass.states`.

boralyl committed:
Assert state of custom project.

MartinHjelmare commented on tests/components/todoist/test_calendar.py:
```diff
@@ -104,3 +104,6 @@ async def test_calendar_custom_project_unique_id(todoist_api, hass, state):
     assert entity is None
+
+    state = hass.states.get("calendar.all_projects")
+    assert "off" == state.state
```
```suggestion
    assert state.state == "off"
```

MartinHjelmare committed:
Adjust assertion style

MartinHjelmare reviewed (approved):
Thanks!
